lib/web/apiserver.go

```diff
@@ -1751,6 +1751,28 @@ func (h *Handler) changeUserAuthentication(w http.ResponseWriter, r *http.Reques
 		return nil, trace.Wrap(err)
 	}
 
+	if h.ClusterFeatures.GetCloud() {
```

Comment:
This block would be much easier to follow if we error out when a function errors:


```go
authPreference, err := ctx.clt.GetAuthPreference(r.Context());
if err != nil {
	return nil, trace.BadRequest("couldn't get auth preferences")
}

if authPreference.GetConnectorName() != constants.PasswordlessConnector {
	users, err := h.cfg.ProxyClient.GetUsers(false)
	if len(users) !== 1 {
		// return err
	}

	authPreference.SetConnectorName(constants.PasswordlessConnector)

	err = ctx.clt.SetAuthPreference(r.Context(), authPreference)

	if err != nil {
		// return err
	}
}
```

If we do not want to do that because of business rules, handling the `err != nil` case first should help.

```go
authPreference, err := ctx.clt.GetAuthPreference(r.Context())
if err != nil {
  // log err
} else {
  users, err := h.cfg.ProxyClient.GetUsers(false)
  if err != nil {
    // log err
  } else {
    // ...
}
  
}

```

Reply:
We don't want to return when something goes wrong here. So we can't return early. I have updated the code https://github.com/gravitational/teleport/pull/17834/files/09ebdb03956031dfb87339c93963fb0207be9003..ac29535915ce67596260c7490b721613742b4906 Let me know what you think